File: solutions/kotlin/flower-field/1/src/main/kotlin/FlowerField.kt
data class FlowerFieldBoard(val garden: List<String>) {


    fun withNumbers(): List<String> {        
        if(garden.size == 0) 
            return listOf<String>()
        var output = mutableListOf<String>()
        val rows = garden.size
        val cols = garden[0].length
        for(y in 0..rows-1){
            var cur_row = ""
            for(x in 0..cols-1){
                if (garden[y][x]=='*')
                    cur_row += '*'
                else {
                    var count=0
                    for(dx in -1..1){
                        for(dy in -1..1){
                            var new_x = x + dx
                            var new_y = y + dy 
                            if(new_y >= 0 && new_y < rows 
                            && new_x >= 0 && new_x < cols
                            && '*' == garden[new_y][new_x])
                                count++;
                        }}
                    if(count>0)
                        cur_row += count.toString()
                    else    
                        cur_row += ' '
                }
            }
            output.add(cur_row)
        }
        return output
    }
}

File: solutions/javascript/flower-field/1/flower-field.js
//
const deltas = [[1,0], [1,1], [0,1], [-1,1], [-1,0], [-1,-1], [0,-1], [1,-1]]

export const annotate = (input) => {  
  
  const numrows = input.length
  if (0 == numrows)
    return [];
  const numcols = input[0].length
  if (0==numcols)
    return input;
  let result = []
  for(let x=0; x<numrows; x++){
    let currow = ""
    for(let y=0; y<numcols; y++){        
        if(input[x][y]==' '){
          let count = 0
          for(let i=0; i<8; i++){
            let newx = x + deltas[i][0]
            let newy = y + deltas[i][1]
              if(0<=newx && newx<numrows 
                 && 0<=newy && newy<numcols){
                count += input[newx][newy] == '*' ? 1 : 0
              }            
          }
          let sum = count>0 ? count.toString() : ' '
          currow += sum
        } else if(input[x][y] == '*') {
          currow += '*'
        }      
    }
    result.push(currow)
  }
  return result
};
